---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# IntegraTRN

---

<!-- badges: start -->
[![License: GPL (>= 3)](https://img.shields.io/badge/License-GPL%20%28%3E%3D%203%29-blue.svg)](https://choosealicense.com/licenses/gpl-3.0/)
![GitHub issues](https://img.shields.io/github/issues/j-y26/IntegraTRN)
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

## Description

The R package `IntegraTRN` integrates transcriptomic, small RNAomic, proteomic, and epigenomic data to construct a transcriptional regulatory network (TRN) underlying the gene expression changes in a developmental or disease (or any continuous/binary) biological context. In particular, the TRN is a network of interacting factors, in which one element can exert a regulatory effect (activating or inhibitory) on the expression of one or more elements. Given the complex nature of transcriptional regulation, the package reveals the key players of such regulation, including transcriptional factors (TFs) and small RNAs. Since gene/protein expression changes are usually the most direct molecular cause of observed phenotypic alterations, the core TRN deciphers changes between two or more conditions and infers the upstream regulatory factors that mediate such changes. The analysis pipeline provided by this package is primarily composed of a two-step process: (1) elucidating the transcriptional and regulatory changes that take place during a biological process, such as development or disease; and (2) performing correlational analysis with rigorous filtering based on biological data to identify the regulatory interactions that are responsible for the observed changes.

The package is developed under the following environment:

- R version 4.3.1 (2023-06-16 ucrt)
- Platform: x86_64-w64-mingw32/x64 (64-bit)
- Running under: Windows 11 x64 (build 22621)

## Installation

To install the latest developmental version of the package:

```{r, eval=FALSE}
require("devtools")
devtools::install_github("j-y26/IntegraTRN", build_vignettes = TRUE)
library("IntegraTRN")
```

To run the shinyApp: `Under construction`

## Overview

```{r, eval=FALSE}
ls("package:IntegraTRN")
data(package = "IntegraTRN")
browseVignettes("IntegraTRN")
```

## Contributions

## References

## Acknowledgements

This package was developed as part of an assessment for 2023
BCB410H: Applied Bioinformatics course at the University of
Toronto, Toronto, CANADA. `IntegraTRN` welcomes issues,
enhancement requests, and other contributions. To submit an issue,
use the GitHub issues.
