---
title: "Integrating multi-omics for constructing transcriptional regulatory networks"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Integrating_multi-omics_for_constructing_transcription_regulatory_networks}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
# Load the package
library(IntegraTRN)
```


## Introduction



## 

### Generate MOList object using multi omics data

```{r message = FALSE, tidy=TRUE}
# Generate MOList object
myMOList <- MOList(
  RNAseq = RNAseq_heart,
  RNAGroupBy = RNAseq_heart_samples$Age,
  smallRNAseq = smallRNAseq_heart,
  smallRNAGroupBy = smallRNAseq_heart_samples$Age
)
```

As long as a MOList object is constructed with RNAseq data, additional omics
data can be added to the existing object using the same syntax. Any existing
data can also be replaced by new data using the same syntax.

```{r message = FALSE, tidy=TRUE}
# Before appending ATACseq data, we first define the path to the peak files
atacPeak1Path <- system.file("extdata", "peak1.bed", package = "IntegraTRN")
atacPeak2Path <- system.file("extdata", "peak2.bed", package = "IntegraTRN")

# Here we add the proteomics and ATACseq peaks data to the existing object
myMOList <- MOList(myMOList,
  pathATACpeak1 = atacPeak1Path, pathATACpeak2 = atacPeak2Path
)
```


```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Perform differential analysis
myMOList <- diffOmics(myMOList,
  rnaseqBatch = RNAseq_heart_samples$Batch,
  program = "DESeq2"
)
```

```{r message = FALSE, tidy=TRUE}
# Annotate small RNAs
myMOList <- annotateSmallRNA(myMOList, anno = "human")
```

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 1. Volcano plot of RNAseq data"}
# Plot Volcano plot of RNAseq data
plotVolcanoRNA(myMOList, log2FC = 0.1)
```

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 2. Volcano plot of small RNAseq data"}
# Plot Volcano plot of small RNAseq data
plotVolcanoSmallRNA(myMOList, log2FC = 0)
```

```{r message = FALSE, tidy=TRUE, echo=FALSE}
# PCA plot of small RNAseq data
pcaPlotList <- plotSmallRNAPCAs(myMOList)
```



```{r message = FALSE, tidy=TRUE}
length(pcaPlotList)
```

In this case, we have 6 PCA plots, one for each small RNA type. We can plot
the PCA plot for miRNA using the following command:

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 3. PCA plot of miRNA"}
pcaPlotList$miRNA
```

Or, we can generate a combined PCA plot for all small RNA types using the
following command:

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=9, fig.height=6, fig.cap="Figure 4. Combined PCA plot of all small RNA types"}
# Use the ggpubr package to combine the PCA plots
ggpubr::ggarrange(pcaPlotList$miRNA,
  pcaPlotList$circRNA,
  pcaPlotList$piRNA,
  pcaPlotList$snoRNA,
  pcaPlotList$snRNA,
  pcaPlotList$tRNA,
  ncol = 3, nrow = 2
)
```

```{r message = FALSE, tidy=TRUE, echo=FALSE}
# Further match the RNAseq and small RNAseq data
myMOList <- matchSamplesRNAsmallRNA(myMOList,
  sampleDFRNAseq = RNAseq_heart_samples,
  sampleDFSmallRNAseq = smallRNAseq_heart_samples
)
```

```{r message = FALSE, tidy=TRUE}
exportMatchResult(myMOList) %>% head(5)     # For the first 5 matches
```

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Annotate ATAC Peaks
library("TxDb.Hsapiens.UCSC.hg38.knownGene")
library("org.Hs.eg.db")
library("BSgenome.Hsapiens.UCSC.hg38")
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
annoDb <- "org.Hs.eg.db"
bsgenome <- BSgenome.Hsapiens.UCSC.hg38

# Retrieve position weight matrices (PWMs) from JASPAR


myMOList <- annotateATACPeaksMotif(myMOList,
  tssRegion = c(-3000, 3000),
  TxDb = txdb,
  annoDb = annoDb,
  bsgenome = bsgenome,
  pwmL = jasparVertebratePWM
)
```


```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 5. Annotations on ATACseq peaks"}
# Plot annotation Pie chart of the ATACseq peaks
plotATACAnno(myMOList)
```

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 6. ATACseq coverage"}
# Plot ATACseq Coverage
plotATACCoverage(myMOList)
```

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=8, fig.height=4, fig.cap="Figure 7. Motif enrichment on ATACseq peaks"}
# Plot ATACseq Motif Heatmaps
plotATACMotifHeatmap(myMOList, pValue = 0.01)
```

Now we can construct the small RNA - TF - target gene network.

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Load external interaction data to the MOList object
myMOList <- loadExtInteractions(myMOList, 
                                miR2Genes = miR2Genes,
                                tf2Genes = tf2Genes
)
```

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Define the cutoffs for the omics data
omiCutoffs <- setOmicCutoffs(
  rnaAdjPval = 0.05,
  rnaLogFC = 0.1,
  rnaTopGenes = 100,
  smallRNAAdjPval = 0.05,
  smallRNALogFC = 0.1,
  smallRNATopGenes = 100,
  atacMotifPval = 0.01
)
```

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Construct the network
myTRNet <- constructTRN(myMOList, 
                        omiCutoffs = omiCutoffs,
                        smallRNAtypes = "all",
                        targetDirection = "both",
                        predicted = TRUE)
```

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Plot the network
plotNetwork(myTRNet, 
            vertexSize - 5,
            edgeArrowSize = 0.5)

g <- myTRNet@network

igraph::plot.igraph(g, 
                    layout = igraph::layout_with_fr(g),
                    vertex.size = 5,
                    edge.arrow.size = 0.5,
                    edge.curved = 0.2,
                    edge.color = "grey",
                    vertex.label.cex = 0.5,
                    vertex.label.color = "black",
                    vertex.label.family = "sans",
                    vertex.label.font = 2,
                    vertex.label.dist = 0.5)

```
