---
title: "Integrating multi-omics for constructing transcriptional regulatory networks"
output: 
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{Integrating multi-omics for constructing transcriptional regulatory networks}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
bibliography: '`r system.file("REFERENCES.bib", package="IntegraTRN")`'
nocite: |
  @chang2020mirnet, @villanueva2019ggplot2, @dplyr
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE, warning=FALSE}
# Load the package
library(IntegraTRN)
library(dplyr)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
# Since devtools does not install the vignette dependencies, we need to install
# these Suggested packages manually
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
if (!requireNamespace("BSgenome.Hsapiens.UCSC.hg38", quietly = TRUE)) {
  BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
}
if (!requireNamespace("TxDb.Hsapiens.UCSC.hg38.knownGene", quietly = TRUE)) {
  BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")
}
if (!requireNamespace("org.Hs.eg.db", quietly = TRUE)) {
  BiocManager::install("org.Hs.eg.db")
}
if (!requireNamespace("ggpubr", quietly = TRUE)) {
  install.packages("ggpubr")
}
if (!requireNamespace("RColorBrewer", quietly = TRUE)) {
  install.packages("RColorBrewer")
}
if (!requireNamespace("ggrepel", quietly = TRUE)) {
  install.packages("ggrepel")
}
```

## Introduction

Understanding the molecular mechanisms underlying complex diseases or during the
course of development has been a major goal in biomedical research. The
regulation of the transcriptome, both at the transcriptional and post-
transcriptional levels, is a key component of these mechanisms. However, the
complexity of the regulatory networks, in particular as it involves multiple
epigenetic players, renders the dissection of these mechanisms challenging. 
In particular, the most common may to construct a transcriptional regulatory
network (TRN) involves using known interactions at a certain condition.
However, this method is insufficient to explain the dynamics of the TRN in terms
of the changes observed between biological conditions. In this vignette, we
demonstrate how to use the **`IntegraTRN`** package to construct a TRN using 
multi-omics data. 

The package **`IntegraTRN`** is designed to integrate differential analysis based 
on multiple omics data and construct a TRN based on both curated and predicted
interactions that underlie the differential expression of genes. In the mean
time, the curation of multi-omics data could be difficult, so the package
provides a flexibility to allow users to decide what data can be integrated
based on their own knowledge. The analysis pipeline can be primarily divided
into two parts: 1) differential analysis of omics data; and 2) construction of 
the TRN. In this vignette, we will demonstrate how to use the package to
construct a TRN using **transcriptomic**, **small RNAomic**, and **chromatin accessibility**
data.

See `help(package = "IntegraTRN")` for more information about the package and
`citation("IntegraTRN")` for how to cite the package. To download the package,
please use the following command:

```{r message = FALSE, eval=FALSE}
require("devtools")
devtools::install_github("j-y26/IntegraTRN", build_vignettes = TRUE)
library("IntegraTRN")
```

To list all available functions in the package, use the following command:

```{r message = FALSE, eval=FALSE}
ls("package:IntegraTRN")
```

## Data preparation

The package has provided datasets for demonstration purposes. The datasets are
fully simulated or partially simulated based on real data [@adar2023integrate]. 
Here is an overview of the basic omics data used in this vignette:

```{r message = FALSE, tidy=TRUE}
data("RNAseq_heart") # RNAseq count matrix
data("RNAseq_heart_samples") # RNAseq sample information
data("smallRNAseq_heart") # small RNAseq count matrix
data("smallRNAseq_heart_samples") # small RNAseq sample information
```

These are RNAseq and small RNAseq data from human fetal heart tissues, as well
as the information about the samples.

Please use `?<dataset>` to see the details of each dataset.

We will now illustrate how to use the package to construct a TRN using these
datasets.

## Differential analysis of omics data

The first step of the analysis pipeline is to perform differential analysis
separately on each omic data. During the differential analysis, we can retrieve
insights into the "changes" that happen during the biological process separately
for each omic data. This exploratory analysis is essential to understand what
alterations in the transcriptome can explain difference in the phenotype of 
interest, and what variations in any of the regulatory players could potentially
contribute to these changes.

### Generate **`MOList`** object using multi omics data

To get started, we first need to generate a **`MOList`** object. The **`MOList`** object
stands for **`multi-omics list`**, which is the core of entire analysis pipeline as
it preserves our input data and the results of the differential analysis in a
centralized object. The **`MOList`** object can be generated using the **`MOList`**
function.

The **`MOList`** constructor has a large flexibility in terms of what input omics
data are available. At the minimum, when constructing a **`MOList`** object for the
first time, we need to provide the RNAseq data and the sample grouping
information. For any count-based omics data (RNAseq, small RNAseq, and 
proteomics), the constructor takes a numeric matrix of counts, where each row
represents a gene or a protein, and each column represents a sample. The 
grouping information can be any vector of the same length as the number of
samples, in the same order to the columns of the count matrix. A continuous
or a binary variable can be used. However, we need to make sure we are supplying
the same grouping criteria for all omics data. Here we create a **`myMOList`**
using only RNAseq and small RNAseq data, using the **`age`** of the samples as 
the grouping criteria for downstream differential analysis.

```{r message = FALSE}
myMOList <- MOList(
  RNAseq = RNAseq_heart,
  RNAGroupBy = RNAseq_heart_samples$Age,
  smallRNAseq = smallRNAseq_heart,
  smallRNAGroupBy = smallRNAseq_heart_samples$Age
)
```

As long as a **`MOList`** object is constructed with RNAseq data, additional omics
data can be added to the existing object using the same syntax. Any existing
data can also be replaced by new data using the same syntax.

Specifically for ATACseq data, we need to provide the path to the peak files
instead of the count matrix. Since most peak callers output peak files in [BED
format](https://genome.ucsc.edu/FAQ/FAQformat.html), we can simply provide a 
path to a BED file, for each condition. Our 
package has already provided two simulated peak files in the **`extdata`** folder.

```{r message = FALSE, tidy=TRUE}
atacPeak1Path <- system.file("extdata", "peak1.bed", package = "IntegraTRN")
atacPeak2Path <- system.file("extdata", "peak2.bed", package = "IntegraTRN")
```

A consensus peak file should be provided for each condition. When biological
replicates are used for the ATACseq experiment, the peak files should be merged
into one and filtered for 
[blacklisted regions](https://github.com/Boyle-Lab/Blacklist) before being 
provided to the constructor [@amemiya2019encode]. 
Here, we simply use the previously created **`myMOList`** object to
append the ATACseq data.

```{r message = FALSE}
myMOList <- MOList(myMOList,
  ATACpeak1 = atacPeak1Path,
  ATACpeak2 = atacPeak2Path
)
```

### Differential analysis of omics data

Once the **`MOList`** object is created, we can perform differential analysis on
each omics data separately. However, since differential analysis on count-based
data is different that that of the epigenomic data, the process is slightly
different.

#### Differential analysis of count-based omics data

For count-based omics data, we can use the **`diffOmics`** function to perform
differential analysis. To simplify the process, the **`diffOmics`** function
essentially performs differential expression analysis on all available count
based omics data in the **`MOList`** object. Since we can specify to use either
**`DESeq2`** or **`edgeR`** for the differential analysis, our underlying assumption
is that the expression data follows a negative binomial distribution 
[@love2014moderated] [@robinson2010edger].
Furthermore, the differential expression analysis utilizes a generalized
linear model (GLM) to allow adjustment on covariates. In particular, the
function allows adjustment on the batch effect, which is a common issue in
next-generation sequencing (NGS) data. 

```{r message = FALSE, warning=FALSE}
myMOList <- diffOmics(myMOList,
  rnaseqBatch = RNAseq_heart_samples$Batch,
  program = "DESeq2"
)
```

The **`batch`** argument follow the same criteria as the grouping information
provided during the construction of the **`MOList`** object. Therefore, this 
argument is not limited to the batch effect, but can be any covariate that
we want to adjust for.

```{r message = FALSE, eval=FALSE}
# NOT run, just illustrate the usage of the batch argument
myMOList <- diffOmics(myMOList,
  smallRnaBatch = smallRNAseq_heart_samples$Sex,
  program = "edgeR"
)
```

##### Exploring differential gene expression analysis

To explore the results of the differential analysis, several visualization
methods are provided. First, to see the distribution of differentially
expressed mRNAs as well as exploring a optimal cutoff for defining genes as
differentially expressed, we can visualize the RNAseq data using a volcano
plot. We can specify the cutoffs for the adjusted p-value and the log2 fold
change (log2FC) to define differentially expressed genes. Here, we define 
`log2FC > 0.1` and `adjusted p-value < 0.05`, which is the default, as the
cutoffs.

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 1. Volcano plot of RNAseq data"}
plotVolcano(myMOList, omic = "RNAseq", log2FC = 0.1)
```

The plot also directly indicates the number of up- and down-regulated genes.

How do we efficiently explore the raw differential expression results? The
**`IntegraTRN`** package has conveniently defined a set of data structures,
primarily S4 classes and methods, that allows users to easily explore the
results. The previously mentioned **`MOList`** object essentially works as a 
list, which contains differential results stored in a **`DETag`** object for
each omics data. This means that we can easily retrieve the differential
expression results via simple subsetting.

```{r message = FALSE, tidy=TRUE}
myRNADETag <- myMOList$DERNAseq
```

A simple snap shot of the **`DETag`** object can be directly invoked.

```{r message = FALSE, tidy=TRUE}
myRNADETag
```

To further obtain the full results of the differential expression analysis,
we can use the **`exportDE`** function. This function returns a data frame
containing the outputs of **`DESeq2`** or **`edgeR`**, with choice of using a 
consistent interface.

```{r message = FALSE, tidy=TRUE}
exportDE(myRNADETag, original = TRUE) %>% head(5)
```

By having **`original = TRUE`**, the data frame remains exactly the same as the
**`DESeq2`** output.

Some times, we may also want to rank the genes to perform gene set enrichment
analysis (GSEA). [@reimand2019pathway] To this end, we can make use of the
subclass of **`DETag`**, which is the **`TOPTag`** object. It automatically
performs piValue transformation [@reimand2019pathway] and ranking based on a 
user-specified criteria. For example, let's retrieve genes that have at least
0.1 log2 fold change and select the top 200  ranked genes.

```{r message = FALSE, warning=FALSE}
myRNATOPTag <- TOPTag(myRNADETag,
  pCutoff = 0.05,
  logFCCutoff = 0.1,
  topGenes = 200,
  direction = "both"
)
```

To see a snap shot:

```{r message = FALSE, tidy=TRUE}
myRNATOPTag
```

Due to the inheritance and object-oriented nature of the S4 class system, we
can handle the **`TOPTag`** object in the same way as the **`DETag`** object.

```{r message = FALSE, tidy=TRUE}
# Export the desired results with rank information
myTopRNAResults <- exportDE(myRNATOPTag)
# Make numeric values easier to read
myTopRNAResults[, 1:4] <- format(myTopRNAResults[, 1:4],
  digits = 4,
  scientific = TRUE
)
head(myTopRNAResults)
```

Here, a rank of 1 and -1 indicates the most up- and down-regulated gene.

For publication purposes, we can also label some genes that we are interested in
on the volcano plot. Let's first retrieve a few differentially expressed genes

```{r message = FALSE}
genesToLabel <- exportDE(myRNADETag) %>%
  dplyr::filter(padj < 0.05, abs(logFC) > 0.1) %>%
  head(3) %>%
  rownames()
```

Then we again plot the volcano plot, with some genes labeled.

We can make use of the **`ggrepel`** package to label the genes on the plot.

```{r message = FALSE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 2. Volcano plot of RNAseq data with some genes labelled"}
rnaVPlot <- plotVolcano(myMOList, omic = "RNAseq", log2FC = 0.1)
rnaVPlot <- rnaVPlot + ggrepel::geom_label_repel(
  data = rnaVPlot$data[rownames(rnaVPlot$data) %in% genesToLabel, ],
  mapping = ggplot2::aes(
    x = logFC,
    y = -log10(padj),
    label = rownames(rnaVPlot$data[rownames(rnaVPlot$data) %in% genesToLabel, ])
  ),
  box.padding = 0.5,
  size = 3
)
rnaVPlot
```


##### Exploring differential small RNA expression analysis

Before moving on to small RNAseq data, we first need to annotate the small
RNAs for their types. To do so, we need to provide a data frame of small RNA
annotation, as in the name of the transcripts for each type of small RNA.
The following is an example:

```{r message = FALSE}
smallRNAAnnotation <- data.frame(
  transcript = c("hsa-miR-1-3p", "hsa-miR-1-5p", "hsa-miR-2-5p"),
  type = c("miRNA", "miRNA", "miRNA")
)
```

To annotate the small RNAseq data, we can use the **`annotateSmallRNA`** function.

```{r message = FALSE, eval=FALSE}
# NOT run, just illustrate the usage of the smallRNAAnnotation argument
myMOList <- annotateSmallRNA(myMOList, anno = smallRNAAnnotation)
```

We need to make sure that the names of the transcripts are consistent with
the names in the count matrix, since different annotation databases may provide
different names for the same transcript.

Alternatively, the package internally provides an annotation database for
human small RNAs. This database is curated from several small RNA annotation
databases, including miRBase, circBase, piRBase, piRNABank, GtRNAdb, GENCODE,
and piRNACluster. Since we are dealing with human data, we can make use of this
internal database.

```{r message = FALSE}
myMOList <- annotateSmallRNA(myMOList, anno = "human")
```

The **`annotateSmallRNA`** function will automatically checks for the coverage of
each small RNA type and will alert the user if the annotation does not cover
all existing transcripts.

Similar to the RNAseq data, we can visualize the distribution of differentially
expressed small RNAs using a volcano plot. 

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 3. Volcano plot of small RNAseq data"}
plotVolcano(myMOList, omic = "smallRNAseq", log2FC = 0)
```

Here, we use the default P-value cutoff, which is `0.05`. In the mean time, the 
separately colors the up and down regulated small RNAs, with the number of each
annotated respectively.

On the other hand, small RNAs are usually a lot more complex and diverse than
mRNAs. Each type of small RNA has its own function and mechanism of action.
Therefore, we can also specifically color each type of small RNAs on the
volcano plot, using a more specialized function. Here, we employ the 
**`RColorBrewer`** package to help with assigning a color palette. Here, we
use the **`BuPu`** palette, which is also by default used by the function.

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 4. Volcano plot of small RNAseq data with each type of small RNA coloured"}
# Use the RColorBrewer package to assign a color palette
library("RColorBrewer")
smallRNAColors <- brewer.pal(6, "BuPu")

# Plot Volcano plot of small RNAseq data
plotVolcanoSmallRNA(myMOList, log2FC = 0, colScheme = smallRNAColors)
```

Furthermore, although the function of microRNAs (miRNAs) is well studied, the
understanding of other types of small RNAs such as piwi-interacting RNAs 
(piRNAs) and small nucleolar RNAs (snoRNAs) is still limited. Therefore, we
can assess the contribution of each type of small RNA to the differences
observed between samples. To do so, we can perform a principal component
analysis (PCA) on each type of small RNA separately.

```{r message = FALSE, tidy=TRUE, results='hide'}
pcaPlotList <- plotSmallRNAPCAs(myMOList)
```

The **`plotSmallRNAPCAs`** function returns a list of PCA plots, one for each type
of small RNA. We can check the length of the list to see how many PCA plots
are generated.

```{r message = FALSE, tidy=TRUE}
length(pcaPlotList)
```

In this case, we have 6 PCA plots, one for each small RNA type. We can plot
the PCA plot for miRNA using the following command:

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 5. PCA plot of miRNA"}
pcaPlotList$miRNA
```

Or, we can generate a combined PCA plot for all small RNA types by making use
of the **`ggpubr`** package.

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=9, fig.height=6, fig.cap="Figure 6. Combined PCA plot of all small RNA types"}
ggpubr::ggarrange(pcaPlotList$miRNA,
  pcaPlotList$circRNA,
  pcaPlotList$piRNA,
  pcaPlotList$snoRNA,
  pcaPlotList$snRNA,
  pcaPlotList$tRNA,
  ncol = 3, nrow = 2
)
```


#### Differential analysis of chromatin accessibility

Analysis of epigenomic data requires several more steps than that of the
count-based omics data. In the case of ATACseq data, genomic regions are
profiled for their accessibility to the hyperactive Tn5 transposase. This
accessibility corresponds to the chromatin state of the region, which can
be either open or closed for DNA binding proteins, particularly transcription
factors (TFs). Therefore, to analyze differential chromatin accessibility, we
need to annotate the peaks with the genes that are nearby. Furthermore, since
we are interested in transcriptional regulation, we can parse through the
open chromatin regions and identify the TF binding motifs that are enriched
in these regions. Differentially enriched motifs can then be used to infer
the TFs that are potentially involved in the regulation of the differentially
expressed genes.

To do so, we first need additional annotation information. Here, we make use of 
the annotation database from the **`TxDb.Hsapiens.UCSC.hg38.knownGene`** package,
which is a **`TxDb`** object that contains the genomic coordinates of all known
genes in the `hg38` genome assembly [@TxDb.Hsapiens.UCSC.hg38.knownGene]. 
We also need to provide a more general
annotation database, which is the **`org.Hs.eg.db`** package to extract the gene
information [@org.Hs.eg.db]. 
Finally, scanning the open chromatin regions for TF binding motifs
requires identifying the sequence of the open chromatin regions. 
Here, we use the corresponding genome FASTA database from the
**`BSgenome.Hsapiens.UCSC.hg38`** package [@BSgenome]. 

```{r message = FALSE, tidy=TRUE, warning=FALSE}
# Annotate ATAC Peaks
library("TxDb.Hsapiens.UCSC.hg38.knownGene")
library("org.Hs.eg.db")
library("BSgenome.Hsapiens.UCSC.hg38")
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
annoDb <- "org.Hs.eg.db"
bsgenome <- BSgenome.Hsapiens.UCSC.hg38
```

Motif enrichment analysis would also require a known set of position weight
matrices (PWMs) for all known curated TFs. The PWM is a matrix that contains
the probability of each nucleotide (A, T, C, G) at every position of the
binding motif. The **`IntegraTRN`** package has already provided a list of PWMs
from the **`JASPAR2022`** database for vertebrates. We can load the PWMs using the
following command:

```{r message = FALSE, tidy=TRUE}
# Load PWMs from JASPAR
data("jasparVertebratePWM")
```

This is a **`PWMatrixList`** object defined in the **`TFBSTools`** package 
[@tan2016tfbstools]. This package
also provides additional functionality to retrieve PWMs from other databases or
tools, such as HOMER etc. Here, we will use our package-provided PWMs from
**`JASPAR2022`** [@castro2022jaspar].

The **`annotateATACPeaksMotif`** function performs both peak annotation and motif
enrichment analysis via a single call. Here, we define the promoter region as
the region **3kb** upstream and **3kb** downstream of the transcription start site.

```{r message = FALSE, warning=FALSE, results='hide'}
myMOList <- annotateATACPeaksMotif(myMOList,
  tssRegion = c(-3000, 3000),
  TxDb = txdb,
  annoDb = annoDb,
  bsgenome = bsgenome,
  pwmL = jasparVertebratePWM
)
```

##### Exploring differential chromatin accessibility

After performing peak annotation and motif enrichment analysis, we can explore
the analysis via several methods.

First, let us take a look at the annotation on the peaks for their genomic
features. This pie chart shows the percentage of peaks that are annotated as
each genomic feature.

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 7. Annotation of ATAC peaks"}
plotATACAnno(myMOList)
```

To further access the quality of the ATACseq experiment, we can plot the
coverage of the ATACseq peaks on the genome. This plot shows the where the peaks
are located in each chromosome.

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=6, fig.height=4, fig.cap="Figure 8. ATACseq coverage"}
plotATACCoverage(myMOList)
```

On the left of the plots shows all differentially enriched motifs.

Beyond globally exploring the ATACseq data with the above plots, we can also
inspect the annotation on each individual peak. Once again, the **`IntegraTRN`**
package utilizes the S4 inheritance system to create a **`PEAKTag`** object
once motif enrichment analysis is performed. This object implements the
generic **`as.data.frame`** method, conveniently allowing us to export the
results as a data frame to see where each individual peaks are located.

```{r message = FALSE, tidy=TRUE}
myMotifs <- as.data.frame(myMOList$DEATAC)
head(myMotifs[, 1:8]) # Many annotations, see the first 8 attributes
```

Then we can explore the key part of the analysis, which is the differentially
enriched motifs. Typically, the differential motif enrichment can be very
stringent especially when the ATACseq libraries are not sequenced very deeply
(less than 100 million reads) [@yan2020reads]. 
Therefore, we can use a more relaxed cutoff for
the p-value. The **`plotATACMotifHeatmap`** function by default utilizes adjusted
p-value, but when a unadjusted p-value is provided, it will use the unadjusted
one and ignore the default. Since our ATACseq data is simulated, here we use
a relaxed cutoff.

```{r message = FALSE, tidy=TRUE, fig.align='center', fig.width=8, fig.height=4, fig.cap="Figure 9. Motif enrichment on ATACseq peaks"}
plotATACMotifHeatmap(myMOList, pValue = 0.01)
```


## Constructing the TRN

Exploring the differential analysis results can provide insights into the
changes that happen during the biological process. In addition, by exploring
these results, have a better understanding of the data and the quality of the
experiment. These understanding should allow us to make better decisions on
deciding the cutoffs used for selecting genes and transcripts for constructing
the TRN.

Now we can move on to the second part of the analysis pipeline, which is to
construct the TRN.

### Match samples between omics data

However, before directly construct the TRN, some data preprocessing is required.
First, the package analyzes predicted interactions by using a tree-based 
algorithm to effectively predict regulatory interactions based on co-expression
between genes and small RNAs [@huynh2010inferring]. 
However, this algorithm requires that the samples
between the RNAseq and small RNAseq data are matched so that co-expression
profiles can be extracted. Therefore, we need to match the samples between the
RNAseq and small RNAseq data [@stuart2011matchit].

We can perform an one-to-one optimal matching between the samples using the
`matchSamplesRNAsmallRNA` function. When we have further information about the 
samples for RNAseq and small RNAseq, we can supply the information as data
frames. The algorithm by default extracts all common variables between the two
data frames and use them for matching. Alternatively, we can also specify the
variables used for matching using the `varMatch` argument. 

Here, we use all common variables, which are `Age` and `Sex` for matching.

```{r message = FALSE, tidy=FALSE, warning=FALSE}
myMOList <- matchSamplesRNAsmallRNA(myMOList,
  sampleDFRNAseq = RNAseq_heart_samples,
  sampleDFSmallRNAseq = smallRNAseq_heart_samples
)
```

Let us take a look at the matched samples.

```{r message = FALSE, tidy=TRUE}
exportMatchResult(myMOList) %>% head(5) # For the first 5 matches
```

There are a total of ``r nrow(exportMatchResult(myMOList))`` matches.

```{r message = FALSE, tidy=TRUE}
nrow(exportMatchResult(myMOList))
```

### Using externally curated interactions

Depending on the biological question, we may want to use externally curated
interactions to be part of the TRN. For example, the miRNet tools provides
a comprehensive database of miRNA-target interactions, but the information
are usually not tissue-specific or all relevant to the biological question 
[@chang2020mirnet].
However, **`IntegraTRN`** provides a flexibility to allow users to import
externally curated interactions, and during the construction of the TRN,
these external interactions are intersected with the predicted interactions
as well as the differential genes/enriched motifs to provide a more focused
TRN.

Let us load some simulated miRNA-target interactions and TF-target interactions.
These data has been provided as part of the package. These interactions are
in the adjacency list format, where it is a list of two vectors, regulator and
target.

```{r message = FALSE, tidy=TRUE}
data("miR2Genes")
data("tf2Genes")
myMOList <- loadExtInteractions(myMOList,
  miR2Genes = miR2Genes,
  tf2Genes = tf2Genes
)
```

### Setting cutoffs for omics data

As mentioned previously, the cutoffs used for selecting genes and transcripts
for constructing the TRN are important. Due to the large number of omics data,
a separate function is provide for users to set the cutoffs. Here we use the
cutoffs defined earlier in the differential analysis.

```{r message = FALSE}
omiCutoffs <- setOmicCutoffs(
  rnaAdjPval = 0.05,
  rnaLogFC = 0.1,
  rnaTopGenes = 50,
  smallRNAAdjPval = 0.05,
  smallRNALogFC = 0.1,
  smallRNATopGenes = 50,
  atacMotifPval = 0.01
)
```

This can also be easily checked.

```{r message = FALSE, tidy=TRUE}
omiCutoffs
```

### Constructing and exploring the TRN

Finally, we can construct the TRN using the **`constructTRN`** function. This 
function contains the logic to deal with different availability of the omics
data as well as the availability of external interactions. We can specify
whether we want to focus only on the up or down regulated genes, or both. We
also need to specify whether we want to use the predicted interactions or not.

Here, we construct a network using both predicted and curated interactions, for
all differentially expressed genes that are both up- and down-regulated. 
When predicted interactions are used, we can make the process faster by
using multiple threads, and a random seed can be specified or reproducibility
[@huynh2010inferring].

```{r message = FALSE, warning=FALSE}
myTRNet <- constructTRN(myMOList,
  omiCutoffs = omiCutoffs,
  smallRNAtypes = "all",
  targetDirection = "both",
  predicted = TRUE,
  nthreads = 1,
  seed = 123
)
```

We can easily plot the network using the **`plotNetwork`** function. We can decide
whether we want an interactive network. For large networks, the interactive
network provides a better way to explore the interactions [@networkd3].

```{r message = FALSE, tidy=TRUE, warning=FALSE, fig.align='center', fig.width=6, fig.height=5, fig.cap="Figure 10. The Transcriptional Regulatory Network"}
plotNetwork(myTRNet, interactive = TRUE)
```

When hovering over the nodes, we can see the name of the element. The each type
of the element (target gene, miRNA, piRNA, TF etc.) are color coded. The layout
of the network is computed by a force-directed algorithm, but we can simply
drag the nodes to better visualize the interactions.

We can also plot a static network using the **`plotNetwork`** function. But this
time, we only plot the up-regulated genes and miRNAs with a more stringent
cutoff.

```{r message = FALSE}
omiCutoffs2 <- omiCutoffs
omiCutoffs2$rnaAdjPval <- 0.01
omiCutoffs2$rnaLogFC <- 0.2
```

Let us construct a new TRN.

```{r message = FALSE, warning=FALSE}
myTRNet2 <- constructTRN(myMOList,
  omiCutoffs = omiCutoffs2,
  smallRNAtypes = "miRNA",
  targetDirection = "up",
  predicted = TRUE
)
```

Then we can plot the network as a static figure [@csardi2006igraph].

```{r message = FALSE, tidy=TRUE, warning=FALSE, fig.align='center', fig.width=6, fig.height=5, fig.cap="Figure 11. The Transcriptional Regulatory Network of Up-regulated Genes"}
plotNetwork(myTRNet2, vertex.size = 10)
```

Some times it is hard to use the default parameters to plot a publication-level
network. However, users can customize the network using the **`igraph`**
package. We can extract the network as an **`igraph`** object and performs 
further customization [@csardi2006igraph].

```{r message = FALSE, tidy=TRUE, warning=FALSE}
exportIgraph(myTRNet2)
```

## References

<div id="refs"></div>

## Session information

```{r message = FALSE, tidy=TRUE}
sessionInfo("IntegraTRN")
```
