---
title: "Optimizing workflows for TRN construction"
output: 
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{Optimizing workflows for TRN construction}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
bibliography: '`r system.file("REFERENCES.bib", package="IntegraTRN")`'
nocite: |
  @chang2020mirnet, @villanueva2019ggplot2, @dplyr
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE, warning=FALSE}
# Load the package
library(IntegraTRN)
library(dplyr)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
# Since devtools does not install the vignette dependencies, we need to install
# these Suggested packages manually
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
if (!requireNamespace("ComplexHeatmap", quietly = TRUE)) {
  BiocManager::install("ComplexHeatmap")
}
if (!requireNamespace("circlize", quietly = TRUE)) {
  install.packages("circlize")
}
if (!requireNamespace("multiMiR", quietly = TRUE)) {
  BiocManager::install("multiMiR")
}
```

## Introduction

Integrating multi-omics data is a powerful approach to understand the complex
regulatory mechanisms that govern biological processes. Unlike most existing
methods that focus on a single type of omics data, incorporating multiple omics
data provides a more solid biological foundation for deciphering the key
regulatory interactions that are involved in the biological process of interest.
However, the analysis of multi-omics data is not trivial. In particular, the
logic behind the workflow given different availability of the omics data 
remains complicated.

Here, in this vignette tutorial, we will demonstrate how to utilize the
functionalities of the **`IntegraTRN`** package to optimize the workflow for
constructing a transcriptional regulatory network (TRN) using multi-omics data.
This vignette will also discusses the logic behind the workflow, informing the
user of their options when constructing a TRN.

We will cover the following topics:

- Customizing the differential analysis of omics data
  
- TRN construction with different data availability

A basic workflow of the **`IntegraTRN`** package has been demonstrated in the
vignette 
`Integrating multi-omics for constructing transcriptional regulatory networks`.
It is highly recommended that users explore the basic workflow first before
moving on to this vignette.

See `help(package = "IntegraTRN")` for more information about the package and
`citation("IntegraTRN")` for how to cite the package.

## Data preparation

The package has provided datasets for demonstration purposes. The datasets are
fully simulated or partially simulated based on real data [@adar2023integrate]. 
Here is an overview of the basic omics data used in this vignette:

```{r message = FALSE, tidy=TRUE}
data("RNAseq_heart") # RNAseq count matrix
data("RNAseq_heart_samples") # RNAseq sample information
data("smallRNAseq_heart") # small RNAseq count matrix
data("smallRNAseq_heart_samples") # small RNAseq sample information
data("protein_heart") # Proteomics count matrix
data("protein_heart_samples") # Proteomics sample information
data("miR2Genes") # miRNA-target interactions, externally curated
data("tf2Genes") # TF-target interactions, externally curated
data("proteinGeneIDConvert") # Protein ID conversion table
```

Please use `?<dataset>` to see the details of each dataset.

ATACseq data are separately provided as peak files. The peak files are in the
BED format, which is a tab-delimited file format for genomic features. The BED
files can be located in the **`extdata`** folder of the package.

```{r message = FALSE, tidy=TRUE}
atacPeak1Path <- system.file("extdata", "peak1.bed", package = "IntegraTRN")
atacPeak2Path <- system.file("extdata", "peak2.bed", package = "IntegraTRN")
```

Let us now begin different ways for optimizing our analytical workflow.

## Customizing the differential analysis of omics data

### Key data structures explained

A key component of the package is the data structures it provides. The following
are a list of the key S4 classes:

- **`MOList`**: An **`MOList`** object is singly responsible for storing all the
  omics data and the results of each step of the analysis. While the raw data
  of each omic type are stored internally in the slots, all analysis results
  are kept as list elements that allow easy subsetting and extraction. By 
  exploiting this feature, we can customize our analysis by extracting and 
  manipulating the results of the analysis.

- **`DETag`**: A **`DETag`** object is the core S4 class that stores the results
  of the differential expression analysis. It provides the users with a
  composite structure for exploring the results of the analysis.

- **`TOPTag`**: A **`TOPTag`** object is a subclass of the **`DETag`** object. 
  Unlike the **`DETag`** object, the **`TOPtag`** object further limits the
  results to a subset of genes that we are interested in. It provides an
  user-friendly interface for filtering and ranking top differentially expressed
  genes/transcripts.

- **`PEAKTag`**: Similar to **`TOPTag`**, a **`PEAKTag`** object is also a 
  subclass of the **`DETag`** object. However, it is specifically designed for
  handling peak annotation and motif enrichment analysis.

- **`TRNet`**: Finally, as the final component of the traditional workflow, a
  **`TRNet`** object is the core S4 class that stores the final TRN. Although 
  itself operates as a data storage object, it can be easily integrated with
  any third-party tools for further topological analysis or customized network
  visualization.

Let us begin with the first step of the analysis pipeline, which is to perform
differential analysis on each omics data separately. We begin by exploring how
these data structures facilitate flexible and detailed analysis of the results.

### Exploring the results of the differential analysis

We will first construct a **`MOList`** object using the data we have prepared.
We consider the age of the samples as the grouping criteria for the differential
analysis.

```{r warning=FALSE}
myMOList <- MOList(
  RNAseq = RNAseq_heart,
  RNAGroupBy = RNAseq_heart_samples$Age,
  smallRNAseq = smallRNAseq_heart,
  smallRNAGroupBy = smallRNAseq_heart_samples$Age,
  proteomics = protein_heart,
  proteomicsGroupBy = protein_heart_samples$Age,
  ATACpeak1 = atacPeak1Path,
  ATACpeak2 = atacPeak2Path
)
```

With the package's one-step differential analysis function, we can easily
explore the results of the differential analysis.

```{r message = FALSE, tidy=TRUE}
myMOList <- diffOmics(myMOList, program = "edgeR")
```

In the **`IntegraTRN`** package, several visualization methods are provided to
explore the results of the differential analysis. These visualization methods
are highly efficient in generating an overview of the results, allowing users to
quickly identify the key changes that happen during the biological process.
However, as biologists who generated the data, we may be interested in some
particular aspects of the results. Therefore, we can further customize our
experience by performing easy extraction of the results.

To obtain the **`DETag`** object for a particular omics data, we can simply
subset the **`MOList`** object using the name of the omics data. For example, to
retrieve differential analysis result of the RNAseq data:

```{r message = FALSE, tidy=TRUE}
myRNADETag <- myMOList$DERNAseq
```

would work just fine. The **`DETag`** class has implemented several methods
for users to easily extract the results. For example, we can directly export
the results as a data frame using the **`exportDE`** function.

```{r message = FALSE, tidy=TRUE}
exportDE(myRNADETag) %>% head(5)
```

Furthermore, in certain conditions, such as generating a heatmap, we may want
to explore only the differentially expressed genes as well as a normalized
count matrix for visualization purposes. Besides subsetting the data frame 
returned by the **`exportDE`** function, we can also further utilize the 
functionality of its subclass, the **`TOPTag`** object. As the name suggests,
the **`TOP`** part implies that we could easily perform filtering to obtain
what we consider to be the "top" differentially expressed genes. For example,
we can retrieve the top 20 differentially expressed genes with any fold change, 
with an adjusted p-value less than 0.05 considered to be significant.

```{r message = FALSE, warning=FALSE}
myRNATOPTag <- TOPTag(myRNADETag,
  pCutoff = 0.05,
  logFCCutoff = 0,
  topGenes = 20,
  direction = "both"
)
```

Similarly, we can also export the results as a data frame. Furthermore, we can
see that a ranking based on **`piValue`** is also provided.

```{r message = FALSE, tidy=TRUE}
exportDE(myRNATOPTag) %>% head(5)
```

In case where we want to obtain a normalized count matrix for visualization,
such as when plotting a heatmap, we can also do so by a simple function call.

```{r message = FALSE, tidy=TRUE}
topRNANormCounts <- exportNormalizedCounts(myRNATOPTag)
```

The **`piValue`** here ranks the genes based on the adjusted p-value and the 
fold change, which is a standard ranking for rank-based enrichment analysis. The 
**`piValue`** is calculated as follows:

$$piValue = -log_{10}(padj) * sign(logFC)$$

where $padj$ is the adjusted p-value and $logFC$ is the log fold change.

We can then make a heatmap using the **`ComplexHeatmap`** package.

```{r message = FALSE, fig.align='center', fig.width=6, fig.height=6, fig.cap="Figure 1. Heatmap of top 20 differentially expressed genes" }
heatmapMatrix <- topRNANormCounts %>%
  as.matrix() %>%
  t() %>%
  scale() %>%
  t()
exprRange <- range(heatmapMatrix)
col <- circlize::colorRamp2(
  c(exprRange[1], 0, exprRange[2]),
  c("blue", "white", "red")
)
ComplexHeatmap::Heatmap(heatmapMatrix,
  col = col,
  show_column_names = FALSE
)
```

Similarly, we can use the list of selected genes to perform gene set enrichment
analysis (GSEA) [@reimand2019pathway]. While a rank based GSEA analysis utilizes
all available genes, we can again utilize the **`TOPTag`** constructor to
perform ranking while not filtering any genes. This can be done by setting a 
non-selective cutoff. Here, a `topGenes = 1` means that we are selecting all
genes that satisfy the cutoffs.

```{r message = FALSE, warning=FALSE}
rankedResult <- TOPTag(myRNADETag,
  pCutoff = 1,
  logFCCutoff = 0,
  topGenes = 1,
  direction = "both"
) %>% exportDE()
```

This can then be used by the users as import to the GSEA analysis.

### Alternative differential analysis methods

For most highly controlled experiments, the differential analysis
between two conditions is fairly straightforward. However, in certain cases,
especially for patient-derived samples, a much higher biological variability
may be observed. How could we handle such variability? With the **`DETag`**
class and the list behavior of the **`MOList`** object, this has been made
possible.

In the following section, we will explore an alternative differential analysis
with multiple covariates. However, any alternative differential expression
analysis can be performed at the user's discretion, and the results can be
easily integrated into the **`MOList`** object.

In the case of a population-based cohort, many attributes could be contributing
to the biological variability. In our example, we can see that the following
attributes are available:

```{r message = FALSE, tidy=TRUE}
colnames(RNAseq_heart_samples)
```

In this case, in the differential expression analysis, we can adjust for both
sex and the batch effect. Again, we use the **`edgeR`** package for the
differential expression analysis.

```{r message = FALSE, warning=FALSE}
# Create the edgeR DGEList object
dgeList <- edgeR::DGEList(
  counts = RNAseq_heart,
  group = RNAseq_heart_samples$Age
)
# Filter lowly expressed genes
keep <- edgeR::filterByExpr(dgeList)
dgeList <- dgeList[keep, ]
# Perform differential expression analysis
design <- model.matrix(~ RNAseq_heart_samples$Age +
  RNAseq_heart_samples$Sex +
  RNAseq_heart_samples$Batch)
dgeList <- edgeR::calcNormFactors(dgeList, method = "TMM") # or a compatible method
dgeList <- edgeR::estimateDisp(dgeList, design)
fit <- edgeR::glmQLFit(dgeList, design) # or other suitable models
qlf <- edgeR::glmQLFTest(fit, coef = ncol(RNAseq_heart_samples))
```

With complex designs, the performance of user-defined differential expression
analysis may be better than the one-step differential expression analysis
provided by the **`IntegraTRN`** package. As we have obtained the results of
the differential expression analysis, integrating the results back into the
**`MOList`** object is fairly straightforward.

To do so, we need to construct a **`DETag`** object. The core component of the
**`DETag`** object is a result data frame, as the original output of **`edgeR`**
or **`DESeq2`**. Additionally is a normalized count matrix.

```{r message = FALSE, tidy=TRUE}
# Obtaining the DE results and normalized count
deResult <- edgeR::topTags(qlf, n = nrow(dgeList)) %>% as.data.frame()
normCounts <- edgeR::cpm(dgeList) %>% as.matrix()

# Construct the DETag object
myNewRNADETag <- DETag(
  DEResult = deResult,
  method = "edgeR",
  normalizedCounts = normCounts
)
```

With the **`DETag`** object, we can then integrate the results back into the
**`MOList`** object. 

```{r message = FALSE, tidy=TRUE}
myMOList$DERNAseq <- myNewRNADETag
```

This would further allow us to perform downstream analysis using the customized
differential expression results.

For the purpose of this specific example, we will not replace the original
analysis results.

While we have demonstrated the flexibility of the **`IntegraTRN`** package in
case of the RNAseq data, other omics data can be handled in the same way. With
these flexibility, we can not only explore the differential analysis results
in a more detailed manner, but also customize the analysis that are specific to
our data and biological question of interest.

## TRN construction with different data availability

Once we have optimize the first part of the analysis pipeline, we can then move
on to the second part, which is to construct the TRN. In this section, we will
discuss how the workflow is impacted by the availability of the omics data,
and how users can make use of their best judgement select the most appropriate
workflow.

### Overview of multi-omics network construction

In a simplistic way, network construction by integrating multiple
omics data is a process of intersecting the regulatory interactions to determine 
the ones that are biologically relevant to the condition of interest. Having more
omics data provided, the analytical workflow of the **`IntegraTRN`** package
could identify the most relevant regulatory interactions with much higher
confidence. However, the availability of the omics data is not always guaranteed.
Therefore, the package take different approaches to handle different scenarios.

In the following sections, we will discuss the contribution of each data type to 
the TRN construction, and how the workflow make use of the data to construct the
TRN. Here, the package focuses on creating a transcription factor - small RNA - mRNA
regulatory network. The following sections describe the role of each omics
data in the TRN.

#### Data preparation for TRN construction

Before we begin, to more properly demonstrate the TRN construction workflow, we 
will downsample the data to reduce the computational time. Here, we randomly
select 500 genes from each of the omics data.

```{r message = FALSE, warning=FALSE}
set.seed(91711)
# Count-based data
RNAseq_heart <- RNAseq_heart[sample(nrow(RNAseq_heart), 500), ]
smallRNAseq_heart <- smallRNAseq_heart[sample(nrow(smallRNAseq_heart), 500), ]
protein_heart <- protein_heart[sample(nrow(protein_heart), 500), ]
# Chromatin accessibility data
peak1 <- read.table(atacPeak1Path, sep = "\t", header = FALSE)
peak2 <- read.table(atacPeak2Path, sep = "\t", header = FALSE)
peak1 <- peak1[sample(nrow(peak1), 500), ]
peak2 <- peak2[sample(nrow(peak2), 500), ]
set.seed(NULL)
```

We will then construct a **`MOList`** object using the downsampled data and
perform differential analysis on each omics data.

```{r message = FALSE, warning=FALSE}
myMOList <- MOList(
  RNAseq = RNAseq_heart,
  RNAGroupBy = RNAseq_heart_samples$Age,
  smallRNAseq = smallRNAseq_heart,
  smallRNAGroupBy = smallRNAseq_heart_samples$Age,
  proteomics = protein_heart,
  proteomicsGroupBy = protein_heart_samples$Age,
  ATACpeak1 = peak1,
  ATACpeak2 = peak2
)
```

```{r message = FALSE, warning=FALSE}
myMOList <- diffOmics(myMOList)
```

We will then begin discussing TRN construction with each component of the
network starting from the targets of the regulatory interactions.

### Gene Expression

The RNAseq data is the most commonly available omics data. In the case of a 
transcriptional regulatory network, the RNAseq data is the most direct
representation of the transcriptional activity. Therefore, the network employs
differentially expressed genes (DEGs) as the regulatory targets of the TRN.

When setting the cutoffs for constructing the TRN, there are several cutoffs to 
consider. These include the fold change, the adjusted p-value, the top number of
genes, and the direction of the change.

```{r message = FALSE}
setOmicCutoffs(
  rnaAdjPval = 0.05,
  rnaLogFC = 0.1,
  rnaTopGenes = 200
)
```

Here, we specifically set the cutoffs for the RNAseq data. While the 
**`setOmicCutoffs`** provides default cutoffs for other omics data, these 
cutoffs are only considered if the selected data types are available.

These parameters are directly correlated to the size of the network, and the
which genes to be included defines the primary biological scope of the network.

We will demonstrate the impact of these cutoffs on the TRN construction once
we start to discuss regulators of the network.

### Protein Expression

Similar to the RNAseq data, the proteomics data is also a direct representation
of the transcriptional activity. However, proteomics provides a much more
biologically relevant representation of the transcriptional activity, since
the protein level remains a downstream product of the transcriptional and
post-transcriptional regulation.

Therefore, the proteomics data are used to further filter the DEGs obtained
from the RNAseq data to extract the phenotypically relevant genes. However, an
interesting limitation of proteomics experiments is that protein coverage is
heavily affected by the sample quality as well as the technology used to 
perform the experiment. Therefore, during prior exploratory analysis, we may
fine-tune the cutoffs for the proteomics data to obtain the most relevant
target gene set while minimizing the loss of information due to technical 
limitations.

For example, when protein coverage is quite high, we could employ a regular 
cutoff. Here, we limit the network to include only target genes that shows
consistent expression changes in both the transcriptomic and proteomic level.

```{r}
setOmicCutoffs(
  rnaAdjPval = 0.05,
  rnaLogFC = 0.1,
  rnaTopGenes = 200,
  proteomicsAdjPval = 0.05,
  proteomicsLogFC = 0
)
```

However, what if our proteomics experiment did not went very well? Are those data
still useful? The answer depends on the biological question of interest. If we
have identified our proteins of interest in the proteomics data, we can still
construct the network using both the RNAseq and proteomics data. However, 
the low coverage may negatively impact the differential analysis. We can either
customize the differential analysis to be more sensitive to the proteomics data,
or we can select genes that show consistent direction of change with the proteins.

```{r}
setOmicCutoffs(
  rnaAdjPval = 0.05,
  rnaLogFC = 0.1,
  rnaTopGenes = 200,
  proteomicsAdjPval = 1,
  proteomicsLogFC = 0
)
```

For integrative use of the proteomics and RNAseq data, we need to make sure that
a correct conversion between the protein IDs and the gene IDs are available. 
This can be done by providing a data frame that contains the conversion table.

```{r message = FALSE}
myMOList <- setGene2Protein(myMOList, proteinGeneIDConvert)
```

Setting the conversion can be done any time before the TRN construction.

Since the TRN is a TF - small RNA - mRNA network, having only the mRNA targets
are unable to obtain any valid regulatory relationships. Therefore, the TRN
construction requires at least one of the other data types to be available.

### Small RNA Expression

One of the key post-transcriptional regulatory mechanisms is the small RNA
regulation. In particular, microRNAs (miRNAs) are the most well-studied small
RNAs. Their regulatory mechanism is well understood, and they are known to be
negatively regulating the expression of their target genes by facilitating
degradation of the target mRNA or inhibiting the translation of the target mRNA.

Small RNA data can come in two forms: small RNAseq data or miRNA-target
interactions that are curated in third-party databases.

#### Small RNAseq

The benefit of small RNAseq data is that it provides a direct representation of
the small RNA expression, which also cover multiple types of small RNAs. Most of
them, except for miRNAs, are not well understood. However, they represent in 
very high abundance in the cell, and may play a role in the regulation of the
transcriptome. Therefore, the small RNAseq data relies on predicted inference of
the small RNA - mRNA interactions. Using the normalized count matrix of the
RNAseq and the small RNAseq data, the package performs optimal pair matching
between the two datasets and utilize a random-forest based classifier to predict
and select the top small RNA - mRNA interactions based on their expression
patterns.

With small RNAseq data, we can leverage all these tiny details to the package.

```{r message = FALSE}
# Annotate the small RNAs
myMOList <- annotateSmallRNA(myMOList, anno = "human")

# Perform optimal matching between the RNAseq and small RNAseq data
myMOList <- matchSamplesRNAsmallRNA(myMOList,
  sampleDFRNAseq = RNAseq_heart_samples,
  sampleDFSmallRNAseq = smallRNAseq_heart_samples
)

# Define a cutoff for RNAseq and small RNAseq
setOmicCutoffs(
  rnaAdjPval = 0.05,
  rnaLogFC = 0,
  rnaTopGenes = 0.2,
  smallRNAAdjPval = 0.05,
  smallRNALogFC = 0,
  smallRNATopGenes = 0.2
)
```

#### miRNA-target interactions

On the other hand, miRNA-target interactions are curated from third-party
databases. Since miRNAs have been well studied, the curated interactions has a
broad coverage of the miRNA - mRNA interactions that have been experimentally
validated. However, these interactions are not necessarily relevant to the
biological process of interest, or not even related to the tissue/cell type of
interest. Therefore, by providing the miRNA-target interactions, the package
utilizes the differential expression analysis results and predicted miRNA
expression to select the most relevant miRNA - mRNA interactions.

While fetching externally curated miRNA-target interactions is not a direct
functionality of the package, it is straightforward as part of the pipeline.
With the differential expression results, we can export a list of names of the
target genes of interest, which are usually the inputs of most database
searches.

```{r message = FALSE, tidy=TRUE}
geneList <- exportDE(myMOList$DERNAseq) %>% rownames()
geneList[1:12]
```

Several databases are available for fetching miRNA-target interactions. For
example, [**`miRNet`**](https://www.mirnet.ca/upload/GeneUploadView.xhtml) 
[@chang2020mirnet] provides a comprehensive list of miRNA-target interactions
with simply fetching using a list of target gene names. Similarly, additional
databases such as [**`miRDB`**](https://mirdb.org/) [@chen2020mirdb] can also
be searched, where users can integrate the results from multiple databases
to obtain a more comprehensive, experimentally validated list of miRNA-target
interactions.

Alternatively, with a similar approach, we can also fetch the miRNA-target
using differentially expressed miRNAs. This can be done by exporting the
differentially expressed miRNAs and use them as input for the database search.

```{r message = FALSE}
smallRNAs <- getDESmallRNA(myMOList,
  padj = 0.05,
  log2fc = 0,
  type = "miRNA"
)
head(smallRNAs, 5)
```

As an example, we use the **`multiMiR`** package [@ru2014multimir] to fetch the
miRNA-target interactions from multiple databases, namely **`miRTarBase`** 
[@huang2020mirtarbase], **`miRecords`** [@xiao2009mirecords], and
**`TarBase`** [@sethupathy2006tarbase].

```{r message = FALSE, warning=FALSE, eval=FALSE}
mir2geneMultiMiR <- multiMiR::get_multimir(
  org = "hsa",
  mirna = rownames(smallRNAs),
  table = "validated",
  summary = FALSE
)
mir2geneMultiMiR <- mir2geneMultiMiR@data
```

To avoid unnecessary database query, here we load an example of the results,
where we randomly selected 500 interactions from the **`multiMiR`** results.

```{r message = FALSE, warning=FALSE}
data("mir2geneMultiMiR")
```

Since in the example data we used the `hsa` nomenclature for small RNAs and
gene symbols for the target genes, we can extract only the information
relevant to us.

```{r message = FALSE, warning=FALSE}
mir2geneMultiMiR <- mir2geneMultiMiR %>%
  dplyr::select(mature_mirna_id, target_symbol) %>%
  dplyr::rename(
    regulator = mature_mirna_id,
    target = target_symbol
  )
```

We can then combine the regulatory interactions identified from the target gene
centric approach and the miRNA centric approach.

```{r message = FALSE, warning=FALSE}
# Combine the two sets of interactions
miR2Genes <- as.data.frame(miR2Genes)
miR2Genes <- rbind(miR2Genes, mir2geneMultiMiR)

# Remove any potential duplications across databases
miR2Genes <- miR2Genes %>%
  dplyr::distinct() %>%
  as.list()
```

After obtaining a comprehensive list of miRNA-target interactions, we can then
use the list to construct the TRN. To do so, we need to load the miRNA-target
interactions into the **`MOList`** object.

```{r message = FALSE}
myMOList <- loadExtInteractions(myMOList,
  miR2Genes = miR2Genes
)
```

#### TRN construction with different levels of biological scope

As we have discussed earlier, the cutoffs for the RNAseq data as well as 
whether to select genes that show consistent expression changes in 
the transcriptomic and proteomic level are important factors that define the
primary biological scope of the network. With the small RNA regulators available,
we can visualize the impact of these cutoffs on the TRN construction.

Let us define a less stringent cutoff first:

```{r message = FALSE}
omiCutoffs <- setOmicCutoffs(
  rnaAdjPval = 0.05,
  rnaLogFC = 0,
  rnaTopGenes = 0.5,
  smallRNAAdjPval = 0.05,
  smallRNALogFC = 0,
  smallRNATopGenes = 0.5
)
```

Let's construct the TRN without considering the protein expression.

```{r message=FALSE, warning=FALSE, include=FALSE}
myTRN1 <- constructTRN(myMOList,
  omiCutoffs = omiCutoffs,
  useOmics = c("RNAseq", "smallRNAseq"),
  targetDirection = "both"
)
```

Without protein, the network we obtained is fairly large and complicated to
interpret.

```{r message = FALSE, warning=FALSE, fig.align='center', fig.width=7, fig.height=5, fig.cap="Figure 2. TRN constructed with without considering protein expression" }
plotNetwork(myTRN1, interactive = TRUE)
```

Alternatively, we can enforce the network to only include genes that show
consistent expression changes in both the transcriptomic and proteomic level.

```{r message = FALSE}
omiCutoffs <- setOmicCutoffs(
  rnaAdjPval = 0.05,
  rnaLogFC = 0.1,
  rnaTopGenes = 0.5,
  smallRNAAdjPval = 0.05,
  smallRNALogFC = 0,
  smallRNATopGenes = 0.5,
  proteomicsAdjPval = 0.05,
  proteomicsLogFC = 0
)
```

The network we obtained is much more stringent in explaining what are the most
important regulations that directly explains the phenotypic changes between 
the biological conditions of interests.

```{r message=FALSE, warning=FALSE, include=FALSE}
myTRN2 <- constructTRN(myMOList,
  omiCutoffs = omiCutoffs,
  useOmics = c("RNAseq", "smallRNAseq", "proteomics"),
  targetDirection = "both"
)
```

Let's see the network.

```{r message = FALSE, warning=FALSE, fig.align='center', fig.width=7, fig.height=5, fig.cap="Figure 3. TRN constructed with coherent protein expression" }
plotNetwork(myTRN2, interactive = TRUE)
```

Besides filtering the target genes, we can also limit our scope to the most
important miRNA - target gene regulation. Since the biological function of
miRNA are known, we could infer further biological insights from a 
miRNA-target gene centric network.

```{r message = FALSE, warning=FALSE, include=FALSE}
myTRN3 <- constructTRN(myMOList,
  omiCutoffs = omiCutoffs,
  useOmics = c("RNAseq", "smallRNAseq", "proteomics"),
  targetDirection = "both",
  smallRNAtypes = "miRNA"
)
```

Let's see the network.

```{r message = FALSE, warning=FALSE, fig.align='center', fig.width=7, fig.height=5, fig.cap="Figure 4. miRNA-target gene TRN constructed with coherent protein expression" }
plotNetwork(myTRN3, interactive = TRUE)
```

Interestingly, we can easily identify modules at this level of biological scope.

### Chromatin Accessibility and Transcription Factor Binding

The chromatin accessibility data is a direct representation of the transcription
factor (TF) binding. Therefore, the chromatin accessibility data is used to
further construct the TRN by selecting the TFs that are found to be
differentially enriched across the conditions of interest.

ATACseq data is used in conjunction with the TF-target interactions to construct
the TRN. Similar to the curated miRNA-target interactions, curated the TF-target 
interactions also face limitations in terms of relevance to the biological 
process of interest. Therefore, the package utilizes the differential 
accessibility analysis with motif enrichment analysis to select the most relevant 
TF - target interactions.

To do so, we need to load the TF-target interactions into the **`MOList`**
object.

```{r message = FALSE}
myMOList <- loadExtInteractions(myMOList,
  tf2Genes = tf2Genes
)
```

The logic behind integrating ATACseq with TF-target interactions is similar to
that of integrating small RNAseq with miRNA-target interactions. However, 
instead of using tree-based classifier to predict the regulatory interactions
between RNAseq and small RNAseq, ATACseq data alone is sufficient to identify
enriched transcription factor motifs. Therefore, with the background of the 
curated TF-target interactions, we can isolate the differentially enriched
motifs from the biological conditions and extract regulatory interactions that
are relevant to the biological process of interest.

### Summary of the TRN construction workflow

In general, the construction of the TRN is a process of intersecting the
regulatory interactions that are determined to be biologically relevant to the
condition of interest. The workflow of the **`IntegraTRN`** package is designed
to be flexible to accommodate different availability of the omics data. However,
with the ease of obtaining externally curated interactions, it is highly 
recommended that users utilize public databases to obtain a comprehensive list
of interactions to establish a global context. Then, the package can be used to
select the most relevant interactions to the biological process of interest.

## References

<div id="refs"></div>

## Session information

```{r message = FALSE, tidy=TRUE}
sessionInfo("IntegraTRN")
```
